compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/inlineclasses/InlineClassManager.kt

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.lower.inlineclasses
+
+import org.jetbrains.kotlin.backend.common.ir.copyTo
+import org.jetbrains.kotlin.backend.common.ir.copyTypeParametersFrom
+import org.jetbrains.kotlin.backend.common.ir.createDispatchReceiverParameter
+import org.jetbrains.kotlin.backend.jvm.JvmLoweredDeclarationOrigin
+import org.jetbrains.kotlin.backend.jvm.lower.inlineclasses.InlineClassAbi.mangledNameFor
+import org.jetbrains.kotlin.ir.builders.declarations.addValueParameter
+import org.jetbrains.kotlin.ir.builders.declarations.buildFun
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.declarations.impl.IrFunctionImpl
+import org.jetbrains.kotlin.ir.symbols.IrSimpleFunctionSymbol
+import org.jetbrains.kotlin.ir.symbols.IrValueParameterSymbol
+import org.jetbrains.kotlin.ir.util.defaultType
+import org.jetbrains.kotlin.ir.util.explicitParameters
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.storage.LockBasedStorageManager
+
+class IrReplacementFunction(
+    val function: IrFunction,
+    val valueParameterMap: Map<IrValueParameterSymbol, IrValueParameter>
+)
+
+/**
+ * Keeps track of replacement functions and inline class box/unbox functions.
+ */
+class InlineClassManager {
+    private val storageManager = LockBasedStorageManager("inline-class-manager")
+
+    /**
+     * Get a replacement for a function or a constructor.
+     */
+    val getReplacementFunction: (IrFunction) -> IrReplacementFunction? =
```

Comment:
I'm having trouble understanding why should anything be lazy in this implementation. I'd suggest maybe using the following approach instead of laziness: first, lower all function declarations to their corresponding replacements, and record the mapping (function -> replacement). Second, traverse the whole file again and lower all usages (calls, references, etc), using the recorded mapping. Basically, `JvmInlineClassLowering.lower` would consist of two passes, with different visitors, with data flowing naturally from the first to the second. What do you think?

Reply:
Sorry, I figured that this approach won't work because we need to replace even usages of functions that are declared outside of our file.